### PR TITLE
feat: adding service account as possible creds

### PR DESCRIPTION
**Pull request by Skarlso**

### CODE REVIEW

File: pkg/ocm/ocm.go [hunk 1]
```diff
@@ -93,13 +94,18 @@ func (c *Client) VerifySourceComponent(ctx context.Context, obj *v1alpha1.Compon
...
 	if err := c.maybeConfigureAccessCredentials(ctx, octx, obj.Spec.Source, obj.Namespace); err != nil {
 		return false, fmt.Errorf("failed to configure credentials for destination: %w", err)
 	}
 
-	repo, err := octx.RepositoryForSpec(ocmreg.NewRepositorySpec(obj.Spec.Source.URL, nil))
+	repoSpec := ocireg.NewRepositorySpec(obj.Spec.Source.URL, nil)
+	repo, err := octx.RepositoryForSpec(repoSpec)
 	if err != nil {
-		return false, fmt.Errorf("failed to get repository for spec: %w", err)
+		panic(err)
```
Comment (Skarlso) [inline pkg/ocm/ocm.go:108]: TODO: remove

File: pkg/ocm/ocm.go [hunk 2]
```diff
@@ -336,3 +290,68 @@ func (c *Client) TransferComponent(ctx context.Context, obj *v1alpha1.ComponentS
...
+		secret := &corev1.Secret{}
+
+		if err := c.client.Get(ctx, types.NamespacedName{
+			Name:      imagePullSecret.Name,
+			Namespace: obj.Namespace,
+		}, secret); err != nil {
+			return fmt.Errorf("failed to get image pull secret: %w", err)
+		}
+
+		data, ok := secret.Data[".dockerconfigjson"]
```
Comment (Skarlso) [inline pkg/ocm/ocm.go:344]: TODO: Make a `const` for this.

**Comment by Skarlso** [comment]
I also need to add authentication tests... That won't be easy. :D

**Comment by Skarlso** [comment]
It's workiiiing! :) 

```
"namespace": "ocm-system", "name": "componentsubscription-sample", "reconcileID": "18eb09e4-0c5d-41cd-b795-22c7a5144a96", "component-name": "github.com/skarlso/ocm-replication"}
time="2023-04-13T10:53:33Z" level=info msg="transferring version" history= logger=ocm version="\"github.com/skarlso/ocm-replication:0.0.1\""
time="2023-04-13T10:53:36Z" level=info msg="  transferring resources" history= logger=ocm version="\"github.com/skarlso/ocm-replication:0.0.1\""
time="2023-04-13T10:53:37Z" level=info msg="adding resource blob" logger=ocm resource=ocm-replication
time="2023-04-13T10:53:41Z" level=info msg="  transferring sources" history= logger=ocm version="\"github.com/skarlso/ocm-replication:0.0.1\""
time="2023-04-13T10:53:41Z" level=info msg="  transferring references" history= logger=ocm version="\"github.com/skarlso/ocm-replication:0.0.1\""
time="2023-04-13T10:53:41Z" level=info msg="  adding component version" history= logger=ocm version="\"github.com/skarlso/ocm-replication:0.0.1\""
2023-04-13T10:53:51Z	INFO	starting reconcile loop	{"controller": "componentsubscription", "controllerGroup": "delivery.ocm.software", "controllerKind": "ComponentSubscription", "ComponentSubscription": {"name":"componentsubscription-sample","namespace":"ocm-system"}, "namespace": "ocm-system", "name": "componentsubscription-sample", "reconcileID": "c716a2c3-0a9a-4bf3-b884-757a3920b78e", "subscription": {"name":"componentsubscription-sample","namespace":"ocm-system"}}
2023-04-13T10:53:51Z	LEVEL(-4)	configuring service account credentials	{"controller": "componentsubscription", "controllerGroup": "delivery.ocm.software", "controllerKind": "ComponentSubscription", "ComponentSubscription": {"name":"componentsubscription-sample","namespace":"ocm-system"}, "namespace": "ocm-system", "name": "componentsubscription-sample", "reconcileID": "c716a2c3-0a9a-4bf3-b884-757a3920b78e"}
2023-04-13T10:53:51Z	LEVEL(-4)	got service account	{"controller": "componentsubscription", "controllerGroup": "delivery.ocm.software", "controllerKind": "ComponentSubscription", "ComponentSubscription": {"name":"componentsubscription-sample","namespace":"ocm-system"}, "namespace": "ocm-system", "name": "componentsubscription-sample", "reconcileID": "c716a2c3-0a9a-4bf3-b884-757a3920b78e", "name": "oci-creds"}
2023-04-13T10:53:51Z	LEVEL(-4)	got newest version from component	{"controller": "componentsubscription", "controllerGroup": "delivery.ocm.software", "controllerKind": "ComponentSubscription", "ComponentSubscription": {"name":"componentsubscription-sample","namespace":"ocm-system"}, "namespace": "ocm-system", "name": "componentsubscription-sample", "reconcileID": "c716a2c3-0a9a-4bf3-b884-757a3920b78e", "version": "0.0.1"}
2023-04-13T10:53:51Z	INFO	latest version and replicated version are a match and not empty	{"controller": "componentsubscription", "controllerGroup": "delivery.ocm.software", "controllerKind": "ComponentSubscription", "ComponentSubscription": {"name":"componentsubscription-sample","namespace":"ocm-system"}, "namespace": "ocm-system", "name": "componentsubscription-sample", "reconcileID": "c716a2c3-0a9a-4bf3-b884-757a3920b78e"}
2023-04-13T10:53:58Z	INFO	component deleted	{"controller": "componentsubscription", "controllerGroup": "delivery.ocm.software", "controllerKind": "ComponentSubscription", "ComponentSubscription": {"name":"componentsubscription-sample","namespace":"ocm-system"}, "namespace": "ocm-system", "name": "componentsubscription-sample", "reconcileID": "4117c644-450d-4f92-9fc3-81ce4690d0f5"}
2023-04-13T10:54:02Z	INFO	component deleted	{"controller": "componentsubscription", "controllerGroup": "delivery.ocm.software", "controllerKind": "ComponentSubscription", "ComponentSubscription": {"name":"componentsubscription-sample","namespace":"ocm-system"}, "namespace": "ocm-system", "name": "componentsubscription-sample", "reconcileID": "44452815-c1c6-4778-8aa3-56c33051c6c8"}
2023-04-13T10:54:16Z	INFO	starting reconcile loop	{"controller": "componentsubscription", "controllerGroup": "delivery.ocm.software", "controllerKind": "ComponentSubscription", "ComponentSubscription": {"name":"componentsubscription-sample","namespace":"ocm-system"}, "namespace": "ocm-system", "name": "componentsubscription-sample", "reconcileID": "769c61e2-8d72-436f-ba10-a4b2999824a1", "subscription": {"name":"componentsubscription-sample","namespace":"ocm-system"}}
2023-04-13T10:54:16Z	LEVEL(-4)	configuring service account credentials	{"controller": "componentsubscription", "controllerGroup": "delivery.ocm.software", "controllerKind": "ComponentSubscription", "ComponentSubscription": {"name":"componentsubscription-sample","namespace":"ocm-system"}, "namespace": "ocm-system", "name": "componentsubscription-sample", "reconcileID": "769c61e2-8d72-436f-ba10-a4b2999824a1"}
2023-04-13T10:54:16Z	ERROR	Reconciler error	{"controller": "componentsubscription", "controllerGroup": "delivery.ocm.software", "controllerKind": "ComponentSubscription", "ComponentSubscription": {"name":"componentsubscription-sample","namespace":"ocm-system"}, "namespace": "ocm-system", "name": "componentsubscription-sample", "reconcileID": "769c61e2-8d72-436f-ba10-a4b2999824a1", "error": "failed to get latest component version: failed to get component versions: failed to list versions for component: failed to authorize: failed to fetch anonymous token: unexpected status: 401 Unauthorized"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/Users/skarlso/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:329
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/Users/skarlso/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:274
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/Users/skarlso/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.4/pkg/internal/controller/controller.go:235
2023-04-13T10:54:16Z	INFO	starting reconcile loop	{"controller": "componentsubscription", "controllerGroup": "delivery.ocm.software", "controllerKind": "ComponentSubscription", "ComponentSubscription": {"name":"componentsubscription-sample","namespace":"ocm-system"}, "namespace": "ocm-system", "name": "componentsubscription-sample", "reconcileID": "32042d39-8c71-4f1b-b248-f0438665afb0", "subscription": {"name":"componentsubscription-sample","namespace":"ocm-system"}}
```


Previously, when we deleted the usage of the service account it was still authenticated because the context remained configured. Now it's no longer doing that.

File: pkg/ocm/ocm.go [hunk 3]
```diff
@@ -336,3 +292,68 @@ func (c *Client) TransferComponent(ctx context.Context, obj *v1alpha1.ComponentS
...
+
+func (c *Client) maybeConfigureServiceAccountAccess(ctx context.Context, octx ocm.Context, obj *v1alpha1.ComponentSubscription) error {
+	logger := log.FromContext(ctx)
+
+	logger.V(4).Info("configuring service account credentials")
+
+	if obj.Spec.ServiceAccountName == "" {
+		return nil
+	}
+
```
Comment (phoban01) [inline pkg/ocm/ocm.go:325]: could this check be outside the method? then you only need to pass in the service account name and the contexts.

File: pkg/ocm/ocm_test.go [hunk 4]
```diff
@@ -8,41 +8,226 @@ import (
...
+						Semver:    "v0.0.1",
+						Source: v1alpha1.OCMRepository{
+							URL: env.repositoryURL,
+						},
+					},
+				}
+			},
+		},
+		{
+			name: "component access with secret ref",
```
Comment (phoban01) [inline pkg/ocm/ocm_test.go:48]: nice

**Review by phoban01 — APPROVED** [review]
A few small nits; tests are really great.

File: pkg/ocm/ocm.go (hunk 3 above)
Comment (Skarlso) [reply pkg/ocm/ocm.go:325]: I'm also using `obj.Namespace,`. But I can pass that in instead. Would minimize the memory burden.

**Pull request merged**